README.md
![nuxt-delay-hydration](https://repository-images.githubusercontent.com/392525648/e8fad899-c221-48ca-bb3e-bf5bde92bfd0)


<h1 align='center'>nuxt-delay-hydration</h1>

<p align='center'>
Improve your Nuxt.js Google Lighthouse score by delaying hydration ⚡️<br>
</p>

<p align='center'>
<a href='https://www.npmjs.com/package/nuxt-delay-hydration'>
<img src='https://img.shields.io/npm/v/nuxt-delay-hydration?color=0EA5E9&label='>
</a>
<a href='https://github.com/harlan-zw/nuxt-delay-hydration/actions/workflows/test.yml'>
<img src='https://github.com/harlan-zw/nuxt-delay-hydration/actions/workflows/test.yml/badge.svg' >
</a>
</p>


**Note: `@nuxt/bridge` and `nuxt3` are not supported yet.** 

## Features

- ⚡️ Instantly increase your Google Lighthouse score
- 🔥 Reduce your "Blocking Time" by as much as 100%
- 🍃 Pre-configured to minimise user experience issues
- 🔁 (optional) Replay pre-hydration clicks

<br>


<details>
  <summary><b>Motivation</b></summary>
<br>
Hydration is the process of hooking up static HTML with your Nuxt app to provide reactivity, used in SSR and SSG.

This hydration process is expensive, especially with Nuxt 2. Google Lighthouse penalises hydration with a high "Total Blocking Time" and "Time to Interactive".

While this is unavoidable in most apps, for static sites which depend on minimal interactivity, it is possible and safe
to delay the hydration to avoid this penalty.

The current solution for delaying hydration is [vue-lazy-hydration](https://github.com/maoberlehner/vue-lazy-hydration) which works well.
However, it can require a lot of tinkering, may break your HMR and not solve the performance issues. 

In `mount` and `init` mode, this module delays Nuxt core code to squeeze out maximum performance, specifically around large payload executions.

Nuxt Delay Hydration aims to provide optimisations with minimal tinkering, by making certain assumptions on trade-offs.
This module isn't built to replace vue-lazy-hydration, but as an abstraction layer on top of it.
</details>

<br>

<details>
  <summary><b>How this module works</b></summary>
<br>
A promise is injected into your app, the location is based on the mode. The promise is resolved as soon as either of these events has fired:

- an interaction event (mouse move, scroll, click, etc)
- an idle callback with a fixed timeout

The idle CPU time hints to Google that these scripts are not required for your app to run.

For example:
- if a Google bot visits the page and has no interaction, out of the box the hydration won't occur until the browser
  idle callback + 6 seconds
- if a user visits the page and moves their cursor or scrolls, the hydration will be triggered immediately. The chance of interacting with the
  non-hydration app will be minimised

Keep in mind, **this is a hacky solution**. Until Google can recognise which scripts are truly blocking, we'll need to rely on this approach.
</details>

<br>

## Install

⚠️ This module is beta, use with caution.

```bash
yarn add nuxt-delay-hydration
# npm i nuxt-delay-hydration
```

_Requirement: SSR, full-static (SSG) is highly recommended._

<details>
  <summary><b>Further Requirements</b></summary>
<br>
Your scripts should not be required to use your website, this is what we're hinting to Google.

To do that you can ensure:
- Full HTML served on the initial response
- Scripts don't trigger a [CLS](https://web.dev/cls/)
- Scripts shouldn't be required for a user to interact with your site
- Avoid using scripts to set images, will affect the [LCP](https://web.dev/lcp/)

Please [benchmark](#performance-auditing) your app before starting.

This module has been tested on documentation sites, blogs and misc content sites.
</details>

<br>

## Usage

Within your `nuxt.config.js` add the following.

```js
// nuxt.config.js
export default {
  buildModules: [
    'nuxt-delay-hydration',
  ],
}
```

_Note: The module will not run in development unless you have enabled [debug](#debugging)._

## Choosing a mode

By default, no mode is selected, you will need to select how you would the module to work.

*Type:* `init` | `mount`| `manual` | `false`

*Default:* `false`

| Type   |      Description    | Use Case |
|----------|:-------------|------:|
| `false` _default_ |  Disable the module | Testing |
| [init](#init-mode) | Delays Nuxt app creation. All code is delayed including plugins and third-party scripts. |  Zero or minimal plugins/modules. |
| [mount](#mount-mode) _recommended_ | Delays Nuxt after creation and before mounting. Plugins and some third-party scripts will work. |   Minimal non-critical plugins and third-party plugins. |
| [manual](#manual-mode) | Delay is provided by the `DelayHydration` component. Extends `vue-lazy-hydration` |  All other apps |

Regardless of the mode you choose, please read [further optimisations](#further-optimisations).

### Init Mode

Delays hydration before the Nuxt app is created. Your entire app, including plugins, will be delayed. 
This will provide the biggest speed improvements however is the riskiest and may increase
other metrics with delayed network requests.

_Pros:_ Provides the biggest blocking time reduction

_Cons:_ Risky if you have critical third party scripts

_Benchmark:_ ~90-100% reduction

```js
export default {
  delayHydration: {
    mode: 'init'
  }
}
```

### Mount Mode

Delays hydration once your app is created (all plugins and vendor bundle loaded) and is about to be mounted. This delays
your layout and page components.

_Pros:_ Safer and still provides good improvements

_Cons:_ May still break certain layouts if they are js dependent.

_Benchmark:_ ~70% reduction

```js
export default {
  delayHydration: {
    mode: 'mount'
  }
}
```


### Manual Mode

Using the manual mode, you manually specify what part of your app you'd like to delay. Useful for when you need some part of the
page to always hydrate immediately, such as a navigation drawer.

_Pros:_ Safest way to optimise 

_Cons:_ Speed improvement based on usage


```js
export default {
  delayHydration: {
    mode: 'manual'
  }
}
```

#### DelayHydration component

Once you have set the mode, you need to use the component. It's recommended you use the component within
your layout file.

```vue
<template>
<div>
    <my-header />
    <delay-hydration>
        <!-- You must have a single child node as the slot -->
        <div>
            <main>
                <nuxt />
            </main>
            <my-footer />
        </div>
    </delay-hydration>
</div>
</template>
```

This component is a wrapper for a pre-configured [vue-lazy-hydration](https://github.com/maoberlehner/vue-lazy-hydration) component.

If you're using [nuxt/components](https://github.com/nuxt/components) then no import is required. Otherwise, you can import 
the component as:

```js
import { DelayHydration } from 'nuxt-delay-hydration/dist/components'
```

The behaviour of the component should be controlled by the [advanced configuration](#advanced-configuration), however props 
are provided for convenience. 

**Props**

_forever_: `boolean:false` Toggle the hydration delay to last forever

_debug_: `boolean:false` Toggle the debug logging

_replayClick_: `boolean:false` Toggle the click replay

## Guides

### Debugging

<details>
  <summary>Debug mode</summary>
<br>
It's recommended that you do thorough testing on your app with the module before deploying it into production. 

To make sure the module is doing what you expect, there is a `debug` mode, which when enabled will log behaviour in the console.

It might be a good idea to always debug on your local environment, in that instance you could do:

```js
export default {
    delayHydration: {
        debug: process.env.NODE_ENV === 'development',
    },
}
```
</details>
<br>

<details>
  <summary>Delay hydration forever</summary>
<br>
Since the hydration will trigger instantly when you interact with the page, it can be useful
to manually delay the hydration forever so you can test the functionality of your app in its non-hydrated state.

```js
export default {
    delayHydration: {
        forever: true
    },
}
```
</details>
<br>
<details>
  <summary>Visualising the hydration status</summary>
<br>
It can be unclear at times whether your app has been hydrated or not if it's quite static, this can make debugging hard.

To make things easier, there is a component `HydrationStatus` which will tell you what's going on. 

```js
<template>
<div>
    <my-header />
    <delay-hydration>
        <div>
            <!-- Show the hydration status, only for debugging -->
            <hydration-status />
            <main>
                <nuxt />
            </main>
            <my-footer />
        </div>
    </delay-hydration>
</div>
</template>
```

If you're using [nuxt/components](https://github.com/nuxt/components) then no import is required. Otherwise, you can import
the component as:

```js
import { HydrationStatus } from 'nuxt-delay-hydration/dist/components'
```
</details>

### Performance Auditing

<details>
  <summary>How to audit your app</summary>
<br>
It's important to measure the performance changes in this module and any configuration changes you make.

The simplest way to benchmark is to use the Google Lighthouse tool within Google Chrome.

However, due to unpredictable results, it's recommended to run numerous iterations of the audit. To make that
easier this module provides a script to run a performance audit with 10 iterations:
1. Install dependencies: `yarn add -D lighthouse chrome-launcher`
2. Build and start your Nuxt app `nuxt generate && nuxt start`
3. Run the audit script `node ./node_modules/nuxt-delay-hydration/scripts/audit.js`

</details>

### Replaying hydration click

<details>
  <summary>What is this and how to enable</summary>
<br>
One of the issues with delaying hydration is that a user interaction event can occur before your scripts are loaded, leading
to a user having to click on something multiple times for it to do what they expect. Think of a hamburger that is triggered using Javascript, if your
app isn't hydrated then clicking it won't do anything.

The best fix for this is to write your [HTML in a way that doesn't need Javascript](https://css-tricks.com/the-checkbox-hack/) to be interactive. 

However, there are use cases where you need to use Javascript and responding to the first click is important. In those instances, you can enable
the replay of the click.

```js
export default {
    delayHydration: {
        replayClick: true
    },
}
```

This is an experimental configuration, you should test this option yourself before implementing it into your production app.
</details>

### Further Optimisations


<details>
  <summary>Load heavy components async</summary>
<br>
When you load in a heavy component synchronously, the javascript will be bundled in with the main application payload.

This will decrease all of your performance metrics. It's recommended you use async imports for these components.

[Analyze](https://nuxtjs.org/docs/2.x/configuration-glossary/configuration-build#analyze) your components and load the big ones async. If you're using nuxt/components, you can 
easily prefix them with `Lazy` to do so, otherwise, you can use the following syntax.

```js
<script>
export default {
  components: {
    AdSlider: () => import('./AdSlider.vue'),
    ArticleContent: () => import('./ArticleContent.vue'),
    CommentForm: () => import('./CommentForm.vue'),
    ImageSlider: () => import('./ImageSlider.vue'),
  },
};
</script>
```
</details>

## Advanced Configuration

Configuration should be provided on the `delayHydration` key within your Nuxt config.

If you're finding the lab or [field data](https://web.dev/lab-and-field-data-differences/) is not performing, you may want to
tinker with this advanced configuration.

### Event Hydration

**hydrateOnEvents**

- Type: `string[]`
- Default: `[ 'mousemove', 'scroll', 'keydown', 'click', 'touchstart', 'wheel' ]`

Controls which browser events should trigger the hydration to resume. By default, it is quite aggressive to avoid
possible user experience issues.

**replayClick**

- Type: `boolean`
- Default: `false`

If the trigger for hydration was a click, you can replay it. Replaying it will re-execute the event when it is presumed your app is hydrated. 

For example, if a user clicks a hamburger icon and hydration is required to open the menu, it would replay the click once hydrated.

⚠️ This is experimental, use with caution.

**replayClickMaxEventAge**

- Type: `number` (milliseconds)
- Default: `1000`

The replay click event will not run if the hydration takes more than the limit specified.

This limit is designed to avoid possible user experience issues.


### Idle Hydration

**idleCallbackTimeout**

- Type: `number` (milliseconds)
- Default: `7000`

When waiting for an idle callback, it's possible to define a max amount of time to wait in milliseconds. This is
 useful when there are a lot of network requests happening.

**postIdleTimeout**

- Type: `{ mobile: number, desktop: number }` (milliseconds)
- Default: `{ mobile: 6000, desktop: 5000, }`

How many to wait (in milliseconds) after the idle callback before we resume the hydration. This extra timeout is required
to avoid the standard "blocking", we need to provide real idle time to lighthouse.

Mobile should always be higher than desktop as the CPU capacity will generally be a lot less than a desktop.

_Note: The default will likely be customised in the future based on further benchmarking._

### Debugging

**debug**

- Type: `boolean`
- Default: `false`

Log details in the console on when hydration is blocked and when and why it becomes unblocked.

**forever**

- Type: `boolean`
- Default: `false`

Run the delay forever, useful for testing your app in a non-hydrated state.

## Benchmarks

- [Pokemon](https://github.com/harlan-zw/nuxt-delay-hydration/blob/master/test/fixtures/pokemon/benchmarks.md)
- [Countries](https://github.com/harlan-zw/nuxt-delay-hydration/blob/master/test/fixtures/countries/benchmarks.md)
- [Manual](https://github.com/harlan-zw/nuxt-delay-hydration/blob/master/test/fixtures/manual/benchmarks.md)

Live examples

- https://www.odysseytraveller.com/
- https://massivemonster.co/

## Credits

- [Markus Oberlehner](https://github.com/maoberlehner). Pioneer of the lazy hydration in Vue 


## License

MIT License © 2021 [Harlan Wilton](https://github.com/harlan-zw)



CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.3.2](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.3.1...0.3.2)

- fix: support legacy @nuxt/kits [`a23b12c`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/a23b12cc3ea4e642f17683eb49bd0937b6abe669)

#### [0.3.1](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.3.0...0.3.1)

> 23 December 2021

- fix: prefer @nuxt/kit-edge [`3719d35`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/3719d3589d52cbc1d076910c667c07315b2248f1)
- Release 0.3.1 [`784d06c`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/784d06c89ad87d9d370477ac899dd57affff96f4)

#### [0.3.0](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.2.4...0.3.0)

> 23 December 2021

- feat: upgrade to new @nuxt/kit module syntax [`26af848`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/26af848cd1f4f29b1ac3eb4fd75f46944f82c472)
- feat: upgrade to new @nuxt/kit module syntax [`b21fdc4`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/b21fdc4889db354926273d96261ae6bda1aa164b)
- Release 0.3.0 [`3e9e45d`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/3e9e45d26d3e3c034e7887f3a56357dfb8167651)

#### [0.2.4](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.2.3...0.2.4)

> 29 August 2021

- Release 0.2.4 [`4588515`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/45885156c5527c5b00db950a7acd2d7d2ba670f7)
- fix: replayClick in safari promise [`4a077b5`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/4a077b5878f2ecc2b04eef8547f5a900a18919fd)
- chore(doc): fixing typo [`5e6c5b0`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/5e6c5b0e50fbd215e1e88593708575b0aee6a25d)

#### [0.2.3](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.2.2...0.2.3)

> 29 August 2021

- chore: add CHANGELOG.md [`632aadc`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/632aadc477f774f6b6e1acd2cf4be2ded564b1f1)
- fix: Safari bug for replayEvent callback [`dadcfd9`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/dadcfd9fc270d1aa7b4bc1697ed774e0418e41e1)
- chore(doc): make the requirements clearer [`5eb8dff`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/5eb8dffe6c0fcd0415f104a8343a442e3dfb5e76)

#### [0.2.2](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.2.1...0.2.2)

> 29 August 2021

- chore(doc): improve how it works section [`7ffcb0b`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/7ffcb0b0c3476af81671ad8988a0eebca7e4fee1)
- fix: Safari bug in using requestIdleCallback [`305df96`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/305df9623078e4d13c9bfcf31df1d2e6946c6097)
- chore(doc): make the motiviation more compelling [`c61047f`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/c61047f52abf5130a84b01edd4703caaf4fa9f26)

#### [0.2.1](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.2.0...0.2.1)

> 28 August 2021

- chore(fixtures): add vite fixture [`64c8417`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/64c84177d8fded29fb7b6e57db8f4ee97e6a18ad)
- chore(doc): clean up [`1697169`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/16971697c307070a6839b1ef7c888e5744b47e4e)
- chore: clean up doc [`7d968b9`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/7d968b90fb2a70a2b5952ae4b9b2d47b03501b91)

#### [0.2.0](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.1.13...0.2.0)

> 27 August 2021

- feat: options refactoring [`8b56fee`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/8b56feedd1ce47653e675d0863c1d48333043f9e)
- chore(doc): misc guides [`aa0884c`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/aa0884ce3a48a79e015ee326efb43c95e4c2587f)
- fix: misc bugs [`ea410d2`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/ea410d268809fa221052c90a3b489d214f4b1ec1)

#### [0.1.13](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.1.12...0.1.13)

> 27 August 2021

- feat: default mode false, ssr check and component dir fix [`c0d580e`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/c0d580eef1f35075f6bf80500f61c5fa50dc8571)
- Release 0.1.13 [`7d4ad32`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/7d4ad3243049b4efb92d8caf7f38d66240e103e8)

#### [0.1.12](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.1.11...0.1.12)

> 26 August 2021

- fix: components build [`6fd6754`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/6fd6754de27804af5db301f8e46795ba8fbbbc73)
- fix: artifact [`f3a6c13`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/f3a6c13267a40b46bb94eca82b37487c6c0344a9)
- Release 0.1.12 [`d3ea1a7`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/d3ea1a75919c89cc269bc8b485dec1c4570a427c)

#### [0.1.11](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.1.10...0.1.11)

> 26 August 2021

- fix: delayhydration component [`fb3558a`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/fb3558ab9c9bfc4b22135cf0cea280ad1c239ee2)
- chore: clean up [`17c8ba4`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/17c8ba4e26dbf99f2bc15bf5062c4732de9ef980)
- Release 0.1.11 [`c56ef2e`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/c56ef2ef666bada8b3949ac3c656e8db12bde1fb)

#### [0.1.10](https://github.com/harlan-zw/nuxt-delay-hydration/compare/0.1.9...0.1.10)

> 25 August 2021

- fix: default mode to mount [`24779e2`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/24779e2072c698231f4b772594a7ea7f34d3668c)
- fix: dodgy delay hydration method [`7ab6414`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/7ab6414f76a8e39567d31d2617a0f4bdd5424000)
- Release 0.1.10 [`b2bdcb9`](https://github.com/harlan-zw/nuxt-delay-hydration/commit/b2bdcb96769f4117b7ac1f8e990cd4af2616e8c7)


src/index.ts
import { dirname, join, resolve } from 'upath'
import { addTemplate, defineNuxtModule } from '@nuxt/kit'
import type { ModuleOptions } from './interfaces'
import { CONFIG_KEY, MODE_DELAY_APP_INIT, MODE_DELAY_APP_MOUNT, MODE_DELAY_MANUAL, NAME } from './constants'
import templateUtils from './util/template'
import logger from './logger'

const nuxtDelayHydration = defineNuxtModule<ModuleOptions>({
  meta: {
    name: NAME,
    configKey: CONFIG_KEY,
  },
  // support @nuxt/kit legacy
  // @ts-ignore
  configKey: CONFIG_KEY,
  // @ts-ignore
  name: NAME,
  defaults: {
    mode: false,
    hydrateOnEvents: [
      'mousemove',
      'scroll',
      'keydown',
      'click',
      'touchstart',
      'wheel',
    ],
    postIdleTimeout: {
      mobile: 6000,
      desktop: 5000,
    },
    idleCallbackTimeout: 7000,
    forever: false,
    debug: false,
    replayClick: false,
    replayClickMaxEventAge: 1000,
  } as ModuleOptions,
  setup: (config: ModuleOptions, nuxt) => {
    if (!config.mode) {
      logger.info(`\`${NAME}\` mode set to \`${config.mode}\`, disabling module.`)
      return
    }
    if (!nuxt.options.ssr) {
      logger.warn(`\`${NAME}\` will only work for SSR apps, disabling module.`)
      return
    }
    // @ts-ignore
    if (nuxt.options.vite && !nuxt.options.vite?.ssr) {
      logger.warn(`\`${NAME}\` only works with vite with SSR enabled, disabling module.`)
      return
    }
    if (!config.debug && nuxt.options.dev) {
      logger.info(`\`${NAME}\` only runs in dev with \`debug\` enabled, disabling module.`)
      return
    }
    if (config.debug && !nuxt.options.dev)
      logger.warn(`\`${NAME}\` debug enabled in a non-development environment.`)
    if (nuxt.options.target !== 'static')
      logger.warn(`\`${NAME}\` is untested in a non-static mode, use with caution.`)

    nuxt.hook('build:before', () => {
      if (process.env.NODE_ENV !== 'test')
        logger.info(`\`${NAME}\` enabled with \`${config.mode}\` mode ${config.debug ? '[Debug enabled]' : ''}`)
      // enable asyncScripts
      // @ts-ignore
      nuxt.options.render.asyncScripts = true
    })

    const delayHydrationPath = join('hydration', 'hydrationRace.js')
    const replayPointerEventPath = join('hydration', 'replayPointerEvent.js')

    addTemplate({
      src: join(resolve(__dirname, 'runtime', 'template', 'delayHydration.js')),
      fileName: delayHydrationPath,
      options: config,
    })

    if (config.replayClick) {
      addTemplate({
        src: resolve(join(__dirname, 'runtime', 'template', 'replayPointerEvent.js')),
        fileName: replayPointerEventPath,
        options: config,
      })
    }

    /**
     * Extend Nuxt components, add our component directory.
     */
    nuxt.hook('components:dirs', (dirs: {path: string; ignore?: string[]}[]) => {
      dirs.push({
        path: resolve(join(__dirname, 'components')),
        ignore: ['index.js'],
      })
    })

    if (config.mode === MODE_DELAY_MANUAL) {
      addTemplate({
        src: resolve(join(__dirname, 'runtime', 'plugin', 'injectDelayHydrationApi.js')),
        fileName: join('hydration', 'pluginDelayHydration.client.js'),
        options: config,
      })
    }

    const utils = templateUtils({ publishPath: join(dirname(__dirname), '.runtime') })

    if (config.mode === MODE_DELAY_APP_INIT || config.mode === MODE_DELAY_APP_MOUNT) {
      /**
       * Hook into the template builder, inject the hydration delayer module.
       */
      // @ts-ignore
      nuxt.hook('build:templates', ({ templateVars, templatesFiles }) => {
        if (config.mode === MODE_DELAY_APP_MOUNT) {
          // @ts-ignore
          const template = utils.matchTemplate(templatesFiles, 'client')
          if (!template)
            return

          templateVars.delayHydrationPath = delayHydrationPath
          templateVars.replayPointerEventPath = replayPointerEventPath
          templateVars.hydrationConfig = config
          // import statement
          template.injectFileContents(
            join(__dirname, 'runtime', 'templateInjects', 'import.js'),
            'import Vue from \'vue\'',
          )
          // actual delayer
          template.injectFileContents(
            join(__dirname, 'runtime', 'templateInjects', 'delayHydrationRace.js'),
            'async function mountApp (__app) {',
          )
          template.publish()
          return
        }

        if (config.mode === MODE_DELAY_APP_INIT) {
          // @ts-ignore
          const template = utils.matchTemplate(templatesFiles, 'index')
          if (!template)
            return

          templateVars.delayHydrationPath = delayHydrationPath
          templateVars.replayPointerEventPath = replayPointerEventPath
          templateVars.hydrationConfig = config
          // import statement
          template.injectFileContents(
            join(__dirname, 'runtime', 'templateInjects', 'import.js'),
            'import Vue from \'vue\'',
          )
          // actual delayer
          template.injectFileContents(
            join(__dirname, 'runtime', 'templateInjects', 'delayHydrationRace.js'),
            'async function createApp(ssrContext, config = {}) {',
          )
          template.publish()
        }
      })
    }
  },
})

export default nuxtDelayHydration
